Cython/Utility/ImportExport.c

```diff
@@ -522,19 +525,30 @@ static PyTypeObject *__Pyx_ImportType(PyObject *module, const char *module_name,
     py_basicsize = 0;
     if (basicsize == (Py_ssize_t)-1 && PyErr_Occurred())
         goto bad;
+    py_itemsize = PyObject_GetAttrString(result, "__itemsize__");
+    if (!py_itemsize)
+        goto bad;
+    itemsize = PyLong_AsSsize_t(py_itemsize);
+    Py_DECREF(py_itemsize);
+    py_itemsize = 0;
+    if (itemsize == (Py_ssize_t)-1 && PyErr_Occurred())
+        goto bad;
 #endif
-    if ((size_t)basicsize < size) {
+    if ((size_t)(basicsize + itemsize) < size) {
         PyErr_Format(PyExc_ValueError,
             "%.200s.%.200s size changed, may indicate binary incompatibility. "
             "Expected %zd from C header, got %zd from PyObject",
-            module_name, class_name, size, basicsize);
+            module_name, class_name, size, basicsize+itemsize);
         goto bad;
     }
-    if (check_size == __Pyx_ImportType_CheckSize_Error && (size_t)basicsize != size) {
+    // varobjects almost have structs  between basicsize and basicsize + itemsize
+    // but the struct isn't always one of the two limiting values
+    if (check_size == __Pyx_ImportType_CheckSize_Error &&
+            ((size_t)basicsize > size || (size_t)(basicsize + itemsize) < size)) {
```

Comment:
I don't think this assumption is correct. `itemsize` gives the size of one item. If there are more items, then this miscalculates the size.

Reply:
Unless I'm missing the point of the comment, the varobject struct is almost inevitably:

```
struct {
    ...
    ItemType items[0];
};
```
or

```
struct {
    ...
    ItemType items[1];
};
```
or
```
struct {
    ...
    ItemType items[];
};
```
So the calculation of "is the struct size right" is based on there either being 0 or 1 items. Obviously extra space may be allocated at runtime for more items, but doesn't need considering here

Reply:
It definitely can be wrong though - I think there should be an option to completely ignore the size check or to specify an expected size. But I don't think it needs to be part of this change